### ISSUE TITLE
key/value pair for tags

### ISSUE THREAD

**Issue by ksuquix** [Floppy/van_dam#788]
This one is tricky as it would probably cause some re-architecture, but it would be nice if instead of just simple tags (or maybe in addition to), we could have key/value pair of tags, so we can do stuff like:

class:fighter
race:dwarf
weapon:axe
genre:fantasy

Then it opens up more possibilities on searches, etc.  (in combination with #787 it could be very powerful for managing large amounts of models)

**Comment by Floppy** [comment]
Interesting idea. Should be possible, the underlying tag architecture is very flexible, it can certainly support multiple different types of tag. It would be a matter of making it usable in the UI.

**Comment by Floppy** [comment]
Yes, https://github.com/mbleigh/acts-as-taggable-on definitely supports custom tag types defined at runtime. So, we could have a site setting with a list of contexts, then dynamically have different tag editors in the UI for each one. Would those be per-library, do you think, or for the app as a whole? I suspect per-library, as my "tabletop" and "misc" libraries would have totally different categorisations.

**Comment by ksuquix** [comment]
per-library seems reasonable.  it makes it more customizable, and tags are usually not so copious as to be onerous to duplicate in different libraries.

Theoretically, you could have it at both levels, then just collate the two sets together, but that could be somewhat burdensome.

**Comment by ksuquix** [comment]
I've spent some time cogitating on this.   The problem isn't exactly tag entry, that's straightforward, it's mostly about tag display / searchability options.

I've got a few changes in my head I'm thinking about implementing next:
1) Add "required" tag regexes per-library 
2) Add "missing" search mechanism to enforce these...
3) On tag cloud display, group tags with same prefix together (probably using an html/css collapsible tree)

For tag entry at a later time, the regexes could display missing things on the edit page, but I'm thinking that would be a later round.
